### PR TITLE
feat: add sum of digits

### PR DESCRIPTION
**Pull request by iamnambiar**
#### Description of Change
Added a file ``sum_of_digits.cpp`` in the ``math`` folder which computes the sum of the digits of an integer.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] Sort by alphabetical order
- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: math/sum_of_digits.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    return sum;
+}
+
+int main() {
+    int num;
+    std::cin >> num;
+    if (num < 0) {
+        num = -1 * num;
+    }
+    std::cout << sum_of_digits(num);
```
Comment (mishraabhinn) [inline math/sum_of_digits.cpp:26]: please add a test case and self-validation:

- create a function void test()

- use a pre-defined number

> say 199765


- call sum_of_digits(int num) function to illustrate its call

- use assert statements to verify the results are as expected.

This ensures the the program is self-reliant and to high-standards due to self-validation. More tests the better

File: math/sum_of_digits.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/**
+ * A simple C++ Program to find the Sum of Digits of input integer.
```
Comment (mishraabhinn) [inline math/sum_of_digits.cpp:2]: See this file documentation #807. 
Please make documentation like that.

**Comment by mishraabhinn** [comment]
`math/sum_of_digits.cpp:28:  Could not find a newline character at the end of the file.  [whitespace/ending_newline] [5]
Done processing math/sum_of_digits.cpp
Total errors found: 1
Python 3.8.3
1 files were modified.
1 C++ files were modified.
cpplint:
Traceback (most recent call last):
  File "/home/runner/work/_temp/2792d87d-e946-43af-bee1-ef2894074c35.py", line 17, in <module>
    subprocess.run(["cpplint", "--filter=-legal/copyright"] + cpp_files, check=True, text=True)
  File "/opt/hostedtoolcache/Python/3.8.3/x64/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.Cal`

**Use Clang-Format to format the code so it conforms google coding style.**

**Comment by kvedala** [comment]
what is the expected result for the number:
`-123456789`
?
Because the input is a signed integer, how is a negative sign treated?

**Comment by iamnambiar** [comment]
@mishraabhinn Ok, thanks for the review. I'll update that now.

**Comment by mishraabhinn** [comment]
> what is the expected result for the numbers:
> `-12345678`
> `-123456789`
> ?
> Because the input is a signed integer, how is a negative sign treated?

 I think program will fail in such case @kvedala. 
For   `-12345678` answer would be `36` it will neglect -1 and treat this as unsigned integer.

**Comment by iamnambiar** [comment]
@kvedala for negative integers, first the number will be converted to positive and then it will find the sum of digits. Sorry, I add the code in ``main()`` instead of ``sum_of_digits()``. I'll fix that.

**Comment by iamnambiar** [comment]
@mishraabhinn I have fixed the issue, like you said.

File: math/sum_of_digits.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * Test function.
+ */
+void test() {
+    int test_case_1 = sum_of_digits(119765);
+    int test_case_2 = sum_of_digits(-12256);
+    assert(test_case_1 == 29);
+    assert(test_case_2 == 16);
+    }
+
+int main() {
```
Comment (Panquesito7) [inline math/sum_of_digits.cpp:37]: Please add a comment structure like:

```c++
/**
 * Main function
 */
```

File: math/sum_of_digits.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    test();
+    std::cout << "Success.";
```
Comment (mishraabhinn) [inline math/sum_of_digits.cpp:39]: add `'\n';` after,  End Of Line

File: math/sum_of_digits.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+void test() {
+    int test_case_1 = sum_of_digits(119765);
+    int test_case_2 = sum_of_digits(-12256);
+    assert(test_case_1 == 29);
+    assert(test_case_2 == 16);
+    }
+
+int main() {
+    test();
```
Comment (mishraabhinn) [inline math/sum_of_digits.cpp:38]: Can you actually make two different function for testing. Change to  test1() and test2(); and make `void test()` split in two parts for `sum_of_digits(119765);` `sum_of_digits(-12256);`

File: math/sum_of_digits.cpp (hunk 3 above)
Comment (iamnambiar) [reply math/sum_of_digits.cpp:37]: Ok.

Comment (iamnambiar) [reply math/sum_of_digits.cpp:67]: @Panquesito7 it's fixed.

**Comment by iamnambiar** [comment]
Will it take time to approve for merge?  Should I need to add anything else in this code? Or do I need to close the pull request? This is my first pull request and I have no idea how this work.

File: math/sum_of_digits.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+/**
+ * Function for testing the sum_of_digits() function with a
+ * second test case of -12256 and assert statement.
+ */
+void test2() {
+    int test_case_2 = sum_of_digits(-12256);
+    assert(test_case_2 == 16);
+    }
```
Comment (kvedala) [inline math/sum_of_digits.cpp:52]: close braces should end at the same indentation as the start brace line begining.

File: math/sum_of_digits.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+/**
```
Comment (kvedala) [inline math/sum_of_digits.cpp:53]: keep a one line gap between functions

File: math/sum_of_digits.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ * 
+ * \detail
+ * First the algorithm check whether the num is negative or positive,
+ * if it is negative, then we neglect the negative sign.
+ * Next, the algorithm extract the last digit of num by dividing by 10 
+ * and extracting the remainder and this is added to the sum.
+ * The number is then divided by 10 to remove the last digit.
+ * This loop continues until num becomes 0. 
+ */
+int sum_of_digits(int num) {
```
Comment (kvedala) [inline math/sum_of_digits.cpp:23]: this means that the input number must be 32bit signed integer at most. Not an issue, and suffices to explain the algorithm.

Comment (iamnambiar) [reply math/sum_of_digits.cpp:23]: Thank you for the reply. Will fix the formatting issues you mentioned. Should I change the integer type to any other type?

Comment (iamnambiar) [reply math/sum_of_digits.cpp:23]: @kvedala I've updated the code.

File: math/sum_of_digits.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright 2020 @author iamnambiar
+ * 
+ * @file
+ * A C++ Program to find the Sum of Digits of input integer.
```
Comment (kvedala) [inline math/sum_of_digits.cpp:5]:
```suggestion
 * \brief A C++ Program to find the Sum of Digits of input integer.
```

**Comment by iamnambiar** [comment]
Ok. Thanks.

**Pull request merged**